beacon-chain/forkchoice/doubly-linked-tree/forkchoice.go

```diff
@@ -65,17 +68,97 @@ func (f *ForkChoice) Head(
 		return [32]byte{}, errors.Wrap(err, "could not apply proposer boost score")
 	}
 
-	if err := f.store.treeRootNode.applyWeightChanges(ctx); err != nil {
+	if err := f.store.treeRootNode.applyWeightChanges(ctx, f.store.proposerBoostRoot, f.store.previousProposerBoostScore); err != nil {
 		return [32]byte{}, errors.Wrap(err, "could not apply weight changes")
 	}
 
 	jc := f.JustifiedCheckpoint()
 	fc := f.FinalizedCheckpoint()
-	currentEpoch := slots.EpochsSinceGenesis(time.Unix(int64(f.store.genesisTime), 0))
-	if err := f.store.treeRootNode.updateBestDescendant(ctx, jc.Epoch, fc.Epoch, currentEpoch); err != nil {
+	currentSlot := slots.CurrentSlot(f.store.genesisTime)
+	secondsSinceSlotStart, err := slots.SecondsSinceSlotStart(currentSlot, f.store.genesisTime, uint64(time.Now().Unix()))
+	if err != nil {
+		log.WithError(err).Error("could not compute seconds since slot start")
+		secondsSinceSlotStart = 0
+	}
+	if err := f.store.treeRootNode.updateBestDescendant(ctx, jc.Epoch, fc.Epoch, currentSlot, secondsSinceSlotStart, f.store.committeeWeight); err != nil {
 		return [32]byte{}, errors.Wrap(err, "could not update best descendant")
 	}
-	return f.store.head(ctx)
+	safeHeadUpdateErr := f.UpdateSafeHead(ctx)
+	head, err := f.store.head(ctx)
+	return head, goErrors.Join(err, safeHeadUpdateErr)
+}
+
+// UpdateSafeHead updates the safe head in the fork choice store.
+func (f *ForkChoice) UpdateSafeHead(
+	ctx context.Context,
+) error {
+	oldSafeHeadRoot := f.store.safeHeadRoot
+	newSafeHeadRoot, err := f.store.safeHead(ctx)
+	if err != nil {
+		return errors.WithMessage(err, "could not update safe head")
+	}
+
+	// The safe head root has changed.
+	if oldSafeHeadRoot != newSafeHeadRoot {
+		newSafeHeadNode, ok := f.store.nodeByRoot[newSafeHeadRoot]
+		if !ok || newSafeHeadNode == nil {
+			return ErrNilNode
+		}
+		newSafeHeadSlot := newSafeHeadNode.slot
+		currentSlot := slots.CurrentSlot(f.store.genesisTime)
+		secondsSinceSlotStart, err := slots.SecondsSinceSlotStart(currentSlot, f.store.genesisTime, uint64(time.Now().Unix()))
+		if err != nil {
+			log.WithError(err).Error("could not compute seconds since slot start")
+		}
+		log.WithFields(logrus.Fields{
+			"currentSlot":        fmt.Sprintf("%d", currentSlot),
+			"sinceSlotStartTime": fmt.Sprintf("%d", secondsSinceSlotStart),
```

Comment:
`secondsSinceSlotStart` is always 0 for me. Maybe we want the measurement to be in ms?

Reply:
Originally I only needed second-level granularity to check where in the slot we see safe head updates. Happy to update it to ms-level if necessary!